### ISSUE TITLE
crypto/x509: Adding Name Constraints

### ISSUE THREAD

**Issue by vanbroup** [golang/go#15196]
I would likt to request for the adoption of change [3230 ](https://go-review.googlesource.com/#/c/3230/ )which is in code review for a long time.

> This change extents the Name Constraint properties by adding the Excluded property for DNSDomains and both the permitted and excluded properties for EmailDomains, IPAddresses and DirectoryNames as specified for the GeneralName property in RFC5280.
> 
> The selected properties are required to create or validate a fully constrained certificate.
> 
> The change also improves the validation of Name Constraints in general and allows the parsing of certificates that have Name Constraints marked as cirtial.

https://go-review.googlesource.com/#/c/3230/ 

To be able to validate all certificates issued according to the CA / Browser Forum Baseline Requirements the full set of name constraints need to be available in GO.

A recent version of the CA / Browser Forum Baseline Requirements states that Technical Constraints in Subordinate CA Certificates MUST be applied via Name Constraints. To support strong and strict certificate path validation and to allow users to see the actual constraints it's important that GO supports the required Name Constraints.

Section 9.7 of the baseline requirements states:

> "If the Subordinate CA Certificate includes the id-kp-serverAuth extended key usage, then the Subordinate CA Certificate MUST include the Name Constraints X.509v3 extension with constraints on dNSName, iPAddress and DirectoryName as follows:-"

The full requirements can be found on: https://cabforum.org/baseline-requirements-documents/

The DirectoryName is also needed in some Microsoft environments. Forbidding all directory names would enforce domain validated certificates even if all certificates under a specific root are used and issued by the same organisation. Or could enforce all certificates to be issued with the same OV certificate details. Email addresses are not specifically handled by the CABForum because they don't cover client auth or s/mime certificates currently but likely to have the same requirements and make the set of constraints supported complete.

**Comment by bradfitz** [comment]
Assigning to @agl to decide the fate of this. His last comment on CL 3230 was:

> I don't think that there's a need to handle verifying email addresses, nor to process DirectoryNames (other than to accept forbidding all DirectoryNames).

**Comment by vanbroup** [comment]
Not handling directory names is against RFC 5280 section 4.2.1.10:

> Applications conforming to this profile **MUST** be able to process name
>    constraints that are imposed on the **directoryName** name form and
>    **SHOULD** be able to process name constraints that are imposed on the
>    **rfc822Name**, uniformResourceIdentifier, dNSName, and iPAddress name
>    forms.

Handeling these name constraints should have minimal impact as the extension is only processed when actually present in the certificate

**Comment by tadukurow** [comment]
Has there been any more thought regarding this issue? 

We are seeing public CA certificates issued with critical name constraint (per RFC 5280) and Excluded subtrees for IP addresses (per the Baseline requirements https://cabforum.org/baseline-requirements-documents/). 

I think x509 should be able to parse BR-compliant CAs.

**Comment by agl** [comment]
@tadukurow can you paste an example for testing?

I thought that I had done this but I see that I forgot to hit submit on https://go-review.googlesource.com/c/36900/. Also, that only supports it for names so I suspect that I'll want to extend it first.

**Comment by tadukurow** [comment]
Here is an example for a CA which does not allow ips in the san of its leaf certificates as per baseline requirements.
 
`
 X509v3 Name Constraints: critical
                Excluded:
                  IP:0.0.0.0/0.0.0.0
                  IP:0:0:0:0:0:0:0:0/0:0:0:0:0:0:0:0
`


```
-----BEGIN CERTIFICATE-----
MIIFojCCA4qgAwIBAgIJAOMWrsjU/s6/MA0GCSqGSIb3DQEBCwUAMD0xCzAJBgNV
BAYTAkdCMQ8wDQYDVQQHDAZMb25kb24xEDAOBgNVBAoMB0RlbW8gQ0ExCzAJBgNV
BAsMAklUMB4XDTE3MDUwNDE1NDQyM1oXDTM3MDQyOTE1NDQyM1owPTELMAkGA1UE
BhMCR0IxDzANBgNVBAcMBkxvbmRvbjEQMA4GA1UECgwHRGVtbyBDQTELMAkGA1UE
CwwCSVQwggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQCy6709MthSnIg6
jl9+fgSdqdok56ElUVVaWiF0skZy17skS05zkdLXT15Su4Ffqs7kRawO4IHZvNVc
oVGOkz4xnG4UT8w9dhtu10yUBbs/hi4/KWtdUFNvu2xU0vEzxuLjCZNyOmy5+rYB
TfD6RIZ3eaElE5dZv0syZfG3WoVJMCYZt3KL7O0FtXSd6lCqUNE+bcbxEZ6Av4aJ
NMGMKK+3ML1lDwmyDalUnkOlcTkxlUt2mBt4MUHzc3fmjY4GgfjlyjKp1bzrBVU5
sGrGGiZgH/kYvqABsUyaf3ejvg413YB66bAOr1sMvm1HGnqock8NpEYCJmWGDBox
2s6o/nNO1i1+eE11Iyty24wx1Wwx1VMXd08CYDPpdhkkfN9LK+8yK7CpOViOy6/a
Xc2h7W0QHvWXCHGoEPDjGPKhSVzB2YR/+Nxlei9SjiI1cOE0ThTH1ikO3/3+S3yT
KGZUJT8Mlra7IUBGfbwm63WRyEI9rlowhtGnA7bNtRPekiA95TqYmf5ZGYyksCso
wnYMU8Onm0JkDu5smeV+UGuqV5kXRiCOKoxdbGiJXWYZHNXEc1rnnDzN5TMJK3S2
qLh0IpAJtQwhS++Oi0oTD3bjsyF9RPtDFnxYbxQJYGO/ig5WG1MQ97QMcUu94ll6
vcjm8B14OUEkSzOVvxJtsaeldquRWQIDAQABo4GkMIGhMB0GA1UdDgQWBBQBMXxW
PYExW1n/j0gJPuO7idEwQDAfBgNVHSMEGDAWgBQBMXxWPYExW1n/j0gJPuO7idEw
QDAPBgNVHRMBAf8EBTADAQH/MA4GA1UdDwEB/wQEAwIBhjA+BgNVHR4BAf8ENDAy
oTAwCocIAAAAAAAAAAAwIocgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAwDQYJKoZIhvcNAQELBQADggIBAAxNFDaTK70DtoX5NkWpfiNau1yyFGryygqb
cTQAwkm8U+qp3Ktwf59eQfCd6fwq0aZXHbF+oLTuJrF9SWrx7RG5maZjAHXQKC52
x5c6jYoWxhBiWcBGf9nYMTs16PKi0gRr5MLWaGsDWG8hMxZPeDW7S7c8Z1UXEul3
6AyHKMON56lqB7i2/b1VB18eDW8g6Y7fyWvNhJBJx5tgAq+zgQNBaKfmmwP3XoRG
2WZXN28+vX6rgpSG9oCei26A4/BYEp37tC2VMTaJ8rvAWaRr+B3jYes94a25E/q1
Vp1/IZBY8nCMZgsz5tusROI+rQPdZPV8AJVSgQsfPZFmBY9V9vMYkuiuTTtnzNCo
vrVU5sQWau+oOKYz/nsip0A7Bq1ZKhcSs3Bzpn1sh870yLWL1ALOLQf2SGrK/lRK
+jcPSifqcme7kUQqjX2+gSn9j0deM3UMoEEfVNt39jnzmDVlxyFzhszLCZFzE+X4
sKuKSsrSCVZdzqC3SKgWrJLTEJajoI5I3pFf4hM9kuinlijOQXtQjPERWeBiUhEH
BqFeLZOvDMcru4UPmaDe5G3J5vNDzk3jvNwc/w1hvESgb8/aiNZFYqNN0bL1RXAH
7LknjK/MZvCZ+nBvjWQMfLhbn6TZqiRHSo55e3DCiCCtffEdGjiPOUdjWRPd97YE
p0Oon62s
-----END CERTIFICATE-----
```

**Comment by vanbroup** [comment]
Hi @agl,

I see that ExcludedDNSDomains is included in the beta for 1.9, but what about the remaining name constraints?

```
	PermittedEmailDomains       []string
	ExcludedEmailDomains        []string
	PermittedIPAddresses        []net.IPNet
	ExcludedIPAddresses         []net.IPNet
	PermittedDirectoryNames     []pkix.Name
	ExcludedDirectoryNames      []pkix.Name
```

These name constraints are needed to parse publicly trusted CA certs, for example, this publicly trusted CA certificate with **critical** Name Constraints (redacted, I can probably share via a private message if needed).

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            ???
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: C=BE, O=GlobalSign nv-sa, CN=GlobalSign CA for AATL - SHA256 - G2
        Validity
            Not Before: Nov 26 00:00:00 2014 GMT
            Not After : Nov 26 00:00:00 2017 GMT
        Subject: C=US, ST=California, O=???, CN=???
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                      ???
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Key Usage: critical
                Certificate Sign, CRL Sign
            X509v3 Extended Key Usage: 
                TLS Web Client Authentication, E-mail Protection, 1.3.6.1.4.1.311.10.3.12
            X509v3 Basic Constraints: critical
                CA:TRUE, pathlen:0
            X509v3 Subject Key Identifier: 
                ???
            X509v3 Authority Key Identifier: 
                keyid:60:C2:F1:52:3E:AD:8C:13:DC:DB:CA:0E:FA:0A:E6:2A:2C:99:4B:D8

            Authority Information Access: 
                OCSP - URI:http://ocsp2.globalsign.com/gsaatlsha2g2
                CA Issuers - URI:http://secure.globalsign.com/cacert/gsaatlsha2g2.crt

            X509v3 Name Constraints: critical
                Permitted:
                  email:.???.com
                  email:???.com
                  DNS:???.com
                  DirName: C = US, ST = California, O = ???
                  DirName: C = US, O = ???
                  DirName: O = ???
                Excluded:
                  IP:0.0.0.0/0.0.0.0
                  IP:0:0:0:0:0:0:0:0/0:0:0:0:0:0:0:0

            X509v3 CRL Distribution Points: 

                Full Name:
                  URI:http://crl.globalsign.com/gs/gsaatlsha2g2.crl

            X509v3 Certificate Policies: 
                Policy: 1.3.6.1.4.1.4146.1.60
                  CPS: https://www.globalsign.com/repository/
                Policy: 1.3.6.1.4.1.2774.2.100

    Signature Algorithm: sha256WithRSAEncryption

```

**Comment by vanbroup** [comment]
As also required by the Mozilla Root Store Policy:

> If the certificate includes the id-kp-serverAuth extended key usage, then the certificate **MUST be Name Constrained** as described in section 7.1.5 of version 1.3 or later of the Baseline Requirements.
> 
> If the certificate includes the id-kp-emailProtection extended key usage, it **MUST include the Name Constraints X.509v3 extension with constraints on rfc822Name**, with at least one name in permittedSubtrees, each such name having its ownership validated according to section 3.2.2.4 of the Baseline Requirements.

https://www.mozilla.org/en-US/about/governance/policies/security-group/certs/policy/#technically-constrained

**Comment by agl** [comment]
The other types did not make it into 1.9, which is now frozen I'm afraid. Shouldn't be a stretch for 1.10, although we do not support validating against emails or directory names so some thought is required about how that would be exposed. (I.e. don't expose and ignore them, even if critical, since we'll never use them in the stdlib and let people extract the extension if they care?)

**Comment by briansmith** [comment]
@agl Directory names are used to chain certificates together issuer<->subject, so directory name constraints are always relevant. Also the name constraint enforcement is supposed to be done independently of what kind of name is being validated. That is, if there is an email address constraint and that constraint is violated, then the certificate chain should be rejected even if you're validating a TLS server name.

**Comment by agl** [comment]
> That is, if there is an email address constraint and that constraint is violated, then the certificate chain should be rejected 

That's the "word of God" approach to certificate validation. I.e. that once validated, everything in the certificate is valid. Go does not take that approach and always considers validity in a specific context.

**Comment by vanbroup** [comment]
I absolutely don't like the approach of ignoring the critical flag, these constraints and the flag is there with a good reason. While I see your argument that Go does not follow the "word of God", which is fine. Let's instead verify what "God" tells us and do more on top of that where possible. If we start leaving out parts of the standards and validation logic the mess is only getting bigger.

We can't say that the users need to parse the extension them selfs as this would mean that they also have to perform the full chain validation and certificate handling them self which is not even possible in all use cases. Name constraints are a key part of the certificate chain validation.

Instead of exposing all name constraint details in the x509.Certificate struct I can imagine that we would prefer to put the name constraints in its own linked struct like pkix.Name. We could also decide to include verification options in x509.VerifyOptions that let users choose to disable name constraint checking.

Another approach is to add a hook to the certificate validation logic to handle 'unknown' or override/extend known extension handling. This way users can fulfill any type of critial extension (also thinking about OCSP must-staple) and perform full name constraints validation in the certificate chain without the need to rewrite the x509 package.

**Comment by briansmith** [comment]
Hi Adam, you probably already know this, but the design, (lack of) documentation, and code of the `Verify()` and `VerifyHostname()` functions suggest that it's good to do it the way I suggest.

For example, IIUC, `VerifyOptions.DNSName` isn't supposed to be required, but IIUC it *is* required now if and only if there are any “permitted” name constraints, which seems weird.

It also seems reasonable to call `Verify()` once (perhaps with a DNSName) and then call `VerifyHostname()` one or more times with different hostnames, for example if one is trying to implement the HTTP/2 connection coalescing feature. However, `VerifyHostname()` doesn't check the name it is given against the name constraints, IIUC.

I think if you want to continue doing things as you suggest then it might be better to do the name constraint check in `VerifyHostname()` instead of in `isValid()`. Note that `isValid()` calls `VerifyHostname()` when the DNSName option is given.

Sorry if I'm misunderstanding something.

**Comment by agl** [comment]
Brian, you're not misunderstanding. (Although I don't think Paul's argument that context-dependent validation is ok but ignoring these constraints is not, is sound. One implies the other unless we're adding contexts for email and so on.)

When the Go X.509 library was first written it was quite unclear whether context-dependent validations were going to be the norm or not, mostly because of DANE. With a DNSSEC-based system, it's not possible to implement a Voice of God-style validation because the authority is inherently limited.

But, as you note, the API was frozen long ago and isn't well demarcated. And, for a long time, it didn't really matter whether people got it right or not.

I've written a CL previously to switch Go over to Voice of God but it didn't end up landing because it was somewhat large and I was overtaken by other stuff. VOG is less surprising and, after Google decided to buttress X.509 rather than replace it, and with Let's Encrypt eliminating most of the friction for large numbers of people, VOG is looking fairly dominant.

So reviving that effort might be a sensible prereq here. I assume it would make Paul happy because VOG implies checking all constraints up-and-down the chain at verification time.

(Although there's still some wooly thinking happening from what I can see. If intermediates need to have negative constraints for all the common other SAN types, then how do new name types ever get added? By that scheme, it would need every constrained intermediate to be updated before it would be safe. At best, I guess, we can declare that all the existing name types got this wrong but, for new ones, any name constraints implicitly exclude them.)

**Comment by rsc** [comment]
I've heard off-issue from some people working with [SPIFFE](https://spiffe.io/) that need URI-based Name Constraints, and SPIFFE in turn is hopefully going to be written in Go and used by a variety of systems, including gRPC and Istio.

I'm trying to understand the current state here. @agl, it sounds like you are okay with moving to the VOG approach to X.509 validation, and he may still have an old CL doing much of the work. The issue is titled "add Name Constraints", though. How would moving to VOG address that issue, or what new API would be needed?

Thanks.

**Comment by rsc** [comment]
I spoke to @agl, and he confirmed that he believes moving to VOG is the right thing to do. He hopes to do that for Go 1.10, and then adding URI-based Name Constraints would follow. If there are crypto experts who want to help, please let us know.

**Comment by vanbroup** [comment]
@rsc, @agl, my previously submitted patch to implement this can't merge and anymore but I can contribute some coding and/or testing time to fix this patch or to create something new. Have there been any further thoughts about the implementation?

I was thinking that instead of creating a list for 'each' general name type in x509.Certificate it might be cleaner to create one GeneralName interface in crypto/x509/pkix?

**crypto/x509**
```go
type NameConstraints struct {
  Permitted pkix.GeneralName
  Excluded pkix.GeneralName
}

// Permitted verifies if the given GeneralName is allowed to be used according to the 
// name constraint rules. The name must be explicitly be Permitted or no other names 
// of the given type must be listed as permitted or excluded.
func (n *NameConstraints) Permitted(pkix.GeneralName) bool {
  ...
}

type Certificate struct {
  ...
  NameConstraintsCritical bool // if true then the name constraints are marked critical
  NameConstraints NameConstraints
  ...
}
```

NameConstraintsCritical could also be exposed as function Critital() on NameConstraints?

**crypto/x509/pkix**
```go
type GeneralName interface{
  String() string
}

type RFC822Name string
type DNSName string
type DirectoryName pkix.Name
type UniformResourceIdentifier net.URL
type IPAddress net.IP
...
```

**Comment by vanbroup** [comment]
@agl thanks for working on this, can we please include directoryName too?

directoryName is used a key restriction to prevent constrained sub-CAs to issue under a non-validated/allowed subject DN.

**Comment by agl** [comment]
> @agl thanks for working on this, can we please include directoryName too?

Based on internal feedback from people who know more about X.509 than I, directoryName constraints were less useful and so I didn't include them in the first CL. The other thing that needs to get done this cycle is EKU checking at chain building time (in order to match the constraints behaviour). After that, it's possible but November is past approaching.

**Comment by bradfitz** [comment]
@agl, what remains here? Anything for Go 1.10?

**Comment by jeffcpullen** [comment]
@agl Thank you for working on this, based on the milestone change, is it safe to assume this will hit with 1.11?

**Comment by bradfitz** [comment]
@jeffcpullen, that's not a safe assumption. It previously had Milestone Go1.10 and wasn't included in Go 1.10.

**Comment by agl** [comment]
What's in 1.10 is everything except constraints on directory names, which seemed to be less important.

DNS, Email, and IP constraints should all be enforced, and EKUs are also enforced as a tree property. (Glancing over the big table in the [test file](https://tip.golang.org/src/crypto/x509/name_constraints_test.go) should give some idea of what the behaviour will be.)

**Comment by rsc** [comment]
Per @agl's earlier comment, all that's left here is DirectoryName name constraints.
It's unclear whether that will happen or not so changing back to NeedsDecision
and updating milestone to Go 1.12.

**Comment by luizluca** [comment]
I got hit by this issue while using [Prometheus](https://prometheus.io/). It fails to check any TLS service, rendering the software much less useful.

Our internal CA use dirname name constraint (and also ip, email, dns name constraint). Any TLS connection using our CA fails with: "x509: unhandled critical extension", with cert.UnhandledCriticalExtensions as "[2.5.29.30] (Name Constraints)". I wasn't able to identify which name constraint failed as code simply considers the whole extension unhandled. However, checking the code, it looks like it was dirname. After that, it was simple to find this issue.

Please, consider including this improvements ASAP. Any popular TLS implementation does support Directory Name name constraints. It would be a shame to discard any golang products simply because it does not have a couple of lines of code. [RFC5280](https://tools.ietf.org/html/rfc5280#page-40) says "Applications conforming to this profile MUST be able to process name constraints that are imposed on the directoryName name form and SHOULD be able to process name constraints that are imposed on the rfc822Name, uniformResourceIdentifier, dNSName, and iPAddress name forms." golang implements all the optional restrictions ("SHOULD") but missed the required restriction ("MUST").

Now, as this is my first day with golang, would it be possible to monkey patch golang internal code with a fix?

As a suggestion, it would be nice if error message could contain the unhandled extension like: "x509: unhandled critical extension [2.5.29.30]". It would be much better if it could inform somehow which entry was not recognized appending some info to the error message or even adding it to cert.UnhandledCriticalExtensions (or turning it from an array to a map).

**Comment by luizluca** [comment]
Besides the usage of name constraints in internal CA, there are lots of public CA using directory name constraints (https://censys.io/certificates?q=%28%28parsed.extensions.name_constraints.excluded_directory_names%3A*+OR+parsed.extensions.name_constraints.permitted_directory_names%3A*%29%29+AND+tags.raw%3A+%22unexpired%22+).

However, I could not find any certificate that sets name_constraints as critical:

https://censys.io/certificates/help?q=parsed.extensions.name_constraints.critical%3A+true

They might have opted to let it be a "best effort" restriction as there are TLS implementations that does not support all kinds of constraints. Go shouldn't be one of those to be blamed.

**Comment by cwawak** [comment]
Hello, I'm very interested in the support of critical name constraints. I am stuck accessing a system with a cert using a critical name constraint!

**Comment by luizluca** [comment]
@cwawak , it does not seem that go staff thinks that this is important.

If you are building the software and you can patch the go used, I have patches since 1.15 that you can apply and build your own "go" that actually works.

https://github.com/luizluca/go/tree/1.15/nameconstraint
...
https://github.com/luizluca/go/tree/1.18/nameconstraint

Just pick the version you need. That is the best I can do for now.
Those branches are outdated. So, just cherry-pick the last commit and put it on top of an updated go branch you need.

For 1.19, I need to (again) solve the conflicts.

**Comment by vanbroup** [comment]
I still think this should be supported by Go core, but maybe we can also think about a different approach, for example by extending the core implementation with a registrable extension handler. In this case a handler could be registered for each extension object identifier, if registered, that package will be used for verification and/or parsing and prevent failures for unknown critical extensions.

An [Extension](https://pkg.go.dev/crypto/x509/pkix#Extension) interface could provide these optional [Verify](https://pkg.go.dev/crypto/x509#Certificate.Verify) and [Parse](https://pkg.go.dev/crypto/x509#ParseCertificate) functionalities which would be called by the relevant certificate functions or the user directly.

**Comment by thepudds** [comment]
Hi @vanbroup 

> I still think this should be supported by Go core, but maybe we can also think about a different approach, for example by extending the core implementation with a registrable extension handler.

If you are suggesting a change in API, then that would need to go through [the proposal process](https://github.com/golang/proposal#readme). 

If this would not need to go through the proposal process, would you mind very briefly elaborating on why not?

(Sorry, just a quick comment, and I haven't looked at any of the details here).

**Comment by luizluca** [comment]
> Hi @vanbroup
> 
> > I still think this should be supported by Go core, but maybe we can also think about a different approach, for example by extending the core implementation with a registrable extension handler.
> 
> If you are suggesting a change in API, then that would need to go through [the proposal process](https://github.com/golang/proposal#readme).

The suggestion of an API change is only because go is failing to implement parts of RFC5280 like, in this case, the MANDATORY directoryName name constraint [RFC5280](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.10):

>    Applications conforming to this profile **MUST be able to process** name
>    constraints that are imposed on the directoryName name form and
>    SHOULD be able to process name constraints that are imposed on the
>    rfc822Name, uniformResourceIdentifier, dNSName, and iPAddress name
>    forms. 

Even the dNSName is optional, but not the directoryName.

**Comment by vanbroup** [comment]
> However, I could not find any certificate that sets name_constraints as critical:\n\nhttps://censys.io/certificates/help?q=parsed.extensions.name_constraints.critical%3A+true

Currently this search returns at list trusted CA certificates with name constraints critical and a directory name included:

https://search.censys.io/search?resource=certificates&q=%28parsed.extensions.name_constraints.critical%3A+true%29+and+labels%3D%60trusted%60+and+parsed.extensions.name_constraints.permitted_directory_names%3A*+%E2%86%A9+

**Comment by vanbroup** [comment]
> If you are suggesting a change in API, then that would need to go through the proposal process.\n\nIf this would not need to go through the proposal process, would you mind very briefly elaborating on why not?\n\n(Sorry, just a quick comment, and I haven't looked at any of the details here).

This issue has been open for about 7+ years, it followed the same process as for other name constraints implementations. 

This is a **minor modification** to the implementation of name constraints which retains API compatibility. It extends the `x509.Certificate` struct with the permitted and excluded directory trees and validates these according RFC 5280 as supposed to.

Without proper validation of (directory) name constraints the process should fail (with critical name constraints) or would continue otherwise, **putting the user at risk** when the constraints are not fullfilled.

**Comment by dbenoit17** [comment]
Hi Go team, what would be the next steps toward a fix for this issue?  A few comments in a [PR](https://github.com/golang/go/pull/39639) indicate that a decision should be made in the issue tracker before proceeding.  Is there sufficient information in this thread to make that call?  Would it be better to open a separate formal change proposal?

Thanks!

**Comment by a-h** [comment]
I encountered this issue at work, so I created a test suite for it - see https://github.com/a-h/nameconstraints

It demonstrates:

- How to generate CAs with a constraint.
- How Go fails to handle that constraint.
- A workaround in Go, by overriding the validation function.
- That other mainstream programming languages correctly support the constraint.
   - Go (does not pass)
   - Python
   - Node
   - dotnet

I created a set of tests which apply DNS (domain) and dirName (OU) constraints:

Port | Domain | Organisational Unit | Should Pass
-- | -- | -- | --
8443 | Correct | Correct | Yes
8444 | Correct | Incorrect | No
8445 | Incorrect | Correct | No
8446 | Incorrect | Incorrect | No

For each of the languages, here are the results:

## Go

Go fails to handle the valid case by default.

However, I also included an attempt at handling the constraint by using a custom verification function. See https://github.com/a-h/nameconstraints/blob/d6e556c906a169cdfdfbfa0ac49a3f093de7a9c7/client/main.go#L16-L27

![Image](https://github.com/user-attachments/assets/50246296-1955-4217-8b4e-91d903df7c41)

## Python

Python correctly handles all cases. See https://github.com/a-h/nameconstraints/blob/main/client-python/app.py

![Image](https://github.com/user-attachments/assets/baddc9d4-b165-4f43-947d-9fdfae4a936d)

## Node

Node correctly handles all cases. See https://github.com/a-h/nameconstraints/blob/main/client-node/index.js

![Image](https://github.com/user-attachments/assets/93871e9d-eec7-4896-96a4-19c2d8dea1ff)

## dotnet

It looks like .NET correctly handles the dirName constraint, but mishandles the case where both DNS name and OU are not correct. Of course, that could be a bug in my test.

![Image](https://github.com/user-attachments/assets/38946d37-a1b9-43e7-9aa9-dadde85bd60f)

## Summary

To proceed, even though this is clearly an issue, maybe it needs this to be fed into the proposal process. Not because it's a major change to the language or API, but because it's security-affecting, and likely needs review/implementation from specific Go team members.

Maybe we're deadlocked - e.g. @vanbroup is willing to do a CL and work it through, but only if it's got a chance of being merged. So, having an accepted proposal would unblock?

I can raise a proposal if that would help.

**Comment by rolandshoemaker** [comment]
We can add support for directory name constraints, but I think the open question is if the complexity of implementing them is worthwhile. DNs are complicated, and properly implementing constraints on them has a number of edge cases that can cause issues.

Based on a search there are currently 10 publicly trusted web PKI certs with critical directory name constraints, which don't appear to be in particularly widely used hierarchies. This would suggest the effort required to implement this feature may not be worthwhile.

Are DN constraints prevalent enough in private PKIs to necessitate this?

Another option is to just fail if we encounter a critical constraints extension with DNs present. I'm actually somewhat surprised we don't already do this. This could either be done at parsing (more aggressive, would cause some certs that currently parse to fail) or at the time of validation (this would require adding an unexported field to x509.Certificate, which could cause weird problems for people who are assuming all fields are public.)

If we decide we do need this, a proposal would be necessary, as we'd need to add new fields to x509.Certificate to represent the constraints.

**Comment by luizluca** [comment]
> We can add support for directory name constraints, but I think the open question is if the complexity of implementing them is worthwhile. DNs are complicated, and properly implementing constraints on them has a number of edge cases that can cause issues.

I respectfully disagree that this involves "a number of edge cases." This isn't rocket science—it's basic crypto. The task essentially boils down to parsing the certificate, checking if the subject matches the directory name constraint, and accepting or rejecting accordingly.

Please take a look at how other libraries handle this, or even at my proposed implementation here: https://go-review.googlesource.com/c/go/+/238362. I stopped keeping it in sync after four years, but it should be relatively trivial to update to the current codebase. I’m not a Go expert or a crypto guru, but I’ve contributed to cryptographic libraries and have implemented this behavior in a few places.

> Are DN constraints prevalent enough in private PKIs to necessitate this?

Most of the usage of DN constraints is in private PKIs. But because they are, by nature, private, it's difficult to measure their prevalence. That doesn’t mean the feature isn’t needed.

As an anecdote: I actively avoid tools implemented in Go unless they are the only viable option and I have full access to the source. Why? Because patching and rebuilding Go itself, then rebuilding all tools for every update, is a huge effort. That effort easily surpasses the complexity of implementing this feature properly in the first place.

What’s worse is when someone unknowingly relies on a Go-based tool and only later discovers it’s non-compliant because it lacks a mandatory feature defined in [RFC 5280 §4.2.1.10](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.10). If the Go team is not going to implement this, then at the very least, this limitation should be explicitly documented at [pkg.go.dev/crypto/x509](https://pkg.go.dev/crypto/x509). Transparency here would save people a lot of time and frustration.

> Another option is to just fail if we encounter a critical constraints extension with DNs present. I'm actually somewhat surprised we don't already do this. This could either be done at parsing (more aggressive, would cause some certs that currently parse to fail) or at the time of validation (this would require adding an unexported field to x509.Certificate, which could cause weird problems for people who are assuming all fields are public.)

Failing to parse the certificate because of a unkown constraint doesn't help and could even be harmful. RFC 5280 is very clear:

>   (...). The X.509 v3 certificate format
>    also allows communities to define private extensions to carry
>    information unique to those communities.  Each extension in a
>    certificate is designated as either critical or non-critical.  A
>    certificate-using system MUST reject the certificate if it encounters
>    a critical extension it does not recognize or a critical extension
>    that contains information that it cannot process.  A non-critical
>    extension MAY be ignored if it is not recognized, but MUST be
>    processed if it is recognized. 

If the constraint is known, go must check it. If unknown, it should reject the certificate if it is a critical extension or ignore if not. You can’t just reject certificates solely because they contain a non-critical unknown extensions - those might be private extensions that Go is not expected to understand at all.

In summary:
- We already have a proposed implementation that could jump-start the feature.
- The feature is required for full RFC 5280 compliance (please, take a look at it).
- Its absence should be clearly documented.

**Comment by samiponkanenssh** [comment]
I would like to mention one very important real world use case where the dirName name constraint are required:
Using external CA issued client certificates for authenticating to a kubernetes cluster

When configuring api-server to accept client certificates signed by an external CA, you configure that external CA as a trust anchor for api-server. This gives the external CA full access to the kubernetes cluster, since the CA can issue certificates with a subject name `O = system:masters`.

If api-server would support the x.509 dirName name constraints, then the admin could require that the external CA has the name constraint which excludes `O = system:masters` before adding it to the client cert trust anchors.
